Fix for #2732 - Saving group as string? But datatype is double!?

lewie opened the pull request:
Fix for https://github.com/openhab/openhab/issues/2732
and Types extended

watou commented:
Is this change compatible with all existing MySQL tables in use for persistence prior to this change? Thanks!

lewie commented:
@watou: YES, this Update is backwards compatible, if not please let me know.

I have extended the Binding:
- Groups can now be saved with correct Types. 
If no Type is found, searching for Type based on its Children.
Bore it has always used DOUBLE for Groups, this Configuration is only working for Number, Dimmer and Rollershutter Item, the other Items could not be saved or could not be read,  which has caused Errors.
- Rollershutter Item Type had a naming Bug: Changed they where saved as DOUBLE instead TINYINT.
- Added Call (VARCHAR(200)) and Location (VARCHAR(30)) Item Types, they where saved as DOUBLE too, which has caused Errors.
- Changed default Type for new unknown Itemtypes to String, because a String can be converted to its original OH-Type without a Problem and therefore causes no Errors.

Rgds
Helmut

watou commented:
Thanks for the explanation, @lewie, much appreciated.

teichsta commented on bundles/persistence/org.openhab.persistence.mysql/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -112,18 +112,22 @@
+		sqlTypes.put("CONTACTITEM", 	"VARCHAR(6)");
+		sqlTypes.put("DATETIMEITEM", 	"DATETIME");
+		sqlTypes.put("DIMMERITEM", 		"TINYINT");
+		//sqlTypes.put("GROUPITEM", 	"DOUBLE");
```
remove commented code

teichsta commented on bundles/persistence/org.openhab.persistence.mysql/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -153,6 +157,35 @@ public void unsetPersistentStateRestorer(PersistentStateRestorer persistentState
+	private String getItemType(Item i) {
+		Item item = i;		
+		if(i instanceof GroupItem){
+			logger.debug("mySQL: Item is GroupItem, try to find ItemType by BaseItem");
```
please add the itemName to log statement (using parameterised logging) for the sake of readability of the logs (also applies to the subsequent statements) … image you only have the log at hand and try to give advice where the problem could be.

teichsta commented on bundles/persistence/org.openhab.persistence.mysql/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -153,6 +157,35 @@ public void unsetPersistentStateRestorer(PersistentStateRestorer persistentState
+		}
+		String itemType = item.getClass().toString().toUpperCase();
+		//Pointsyntax??
+		itemType = itemType.substring(itemType.lastIndexOf('.') + 1);
```
rather use item.getClass().getSimpleName()

teichsta commented on bundles/persistence/org.openhab.persistence.mysql/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -153,6 +157,35 @@ public void unsetPersistentStateRestorer(PersistentStateRestorer persistentState
+		//Pointsyntax??
+		itemType = itemType.substring(itemType.lastIndexOf('.') + 1);
+		if(sqlTypes.get(itemType) == null)
+			return sqlTypes.get("STRINGITEM");
```
add curly braces for the sake of readability

teichsta commented:
Thanks @lewie please find my review comments inline. Best, Thomas E.-E.

lewie commented:
Thanks @teichsta, review changes Done. Rgds Helmut

teichsta commented on bundles/persistence/org.openhab.persistence.mysql/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -153,14 +156,42 @@ public void unsetPersistentStateRestorer(PersistentStateRestorer persistentState
+		}
+		String itemType = item.getClass().getSimpleName().toUpperCase();
+		//Pointsyntax??
+		itemType = itemType.substring(itemType.lastIndexOf('.') + 1);
```
this line should be superfluous when using the Simplename, would you agree?

lewie replied:
Yes, Thanks! :-)

teichsta commented on bundles/persistence/org.openhab.persistence.mysql/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -153,14 +156,42 @@ public void unsetPersistentStateRestorer(PersistentStateRestorer persistentState
+		//Pointsyntax??
+		itemType = itemType.substring(itemType.lastIndexOf('.') + 1);
+		if(sqlTypes.get(itemType) == null){
+			logger.debug("mySQL: ItemType NOT found! Use STRINGITEM for {}", i.getName());
```
add itemType to search to the statement

teichsta commented on bundles/persistence/org.openhab.persistence.mysql/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -153,14 +156,42 @@ public void unsetPersistentStateRestorer(PersistentStateRestorer persistentState
+				logger.debug("mySQL: Did not find ItemType for {}, 'GroupItem:<ItemType>' is not defined in *.items File. Try to use first Child-Members ItemType now.", i.getName());
+				item = ((GroupItem) i).getMembers().get(0);
+				if(item == null){
+					logger.debug("mySQL: ItemType NOT found! Use STRINGITEM for {}", i.getName());
```
add itemType to search to the statement

lewie replied:
A bit confused, does this make sense? When item is null, from where should I get a ItemType? I haven't one at this time, isn't it?
Debug logs, can be more clear, did you mean this? 
my new suggestion:
```
	private String getItemType(Item i) {
		Item item = i;		
		if(i instanceof GroupItem){
			item = ((GroupItem) i).getBaseItem();
			if(item == null){//if GroupItem:<ItemType> is not defined in *.items using StringType
				logger.debug("mySQL: No GroupItem ItemType found for {}, 'GroupItem:<ItemType>' is not defined in *.items File.", i.getName());
				item = ((GroupItem) i).getMembers().get(0);
				if(item == null){
					logger.debug("mySQL: No ItemType found for first Child-Member of GroupItem {}, use ItemType STRINGITEM ({}) as Fallback", i.getName(), sqlTypes.get("STRINGITEM"));
					return sqlTypes.get("STRINGITEM");
				}
			}
		}
		String itemType = item.getClass().getSimpleName().toUpperCase();
		if(sqlTypes.get(itemType) == null){
			logger.debug("mySQL: No sqlType found for ItemType {}, use ItemType STRINGITEM ({}) as Fallback for {}", itemType, sqlTypes.get("STRINGITEM"), i.getName());
			return sqlTypes.get("STRINGITEM");
		}

		logger.debug("mySQL: Use ItemType {} ({}) for Item {}", itemType, sqlTypes.get(itemType), itemType, i.getName());
		return sqlTypes.get(itemType);
	}
```

teichsta replied:
hmm … in fact the missing BaseItem is the problem not the missing ItemType. What about something like :"Cannot detect ItemType because the GroupItems' base type isn't set …"

teichsta commented:
Hi @lewie, thanks for incorporating my feedback. It seems i've overlooked some minor issues. Would you mind incorporating them as well? Sorry for the inconvenience! Best, Thomas E.-E.

lewie commented:
Hi @teichsta, code review changes done.

teichsta commented:
Thanks @lewie!
